torch/_inductor/kernel/conv.py

```diff
@@ -0,0 +1,400 @@
+def convolution(
+    x: TensorBox,
+    weight: TensorBox,
+    bias: TensorBox,
+    stride: List[int],
+    padding: List[int],
+    dilation: List[int],
+    transposed: bool,
+    output_padding: List[int],
+    groups: int,
+):
+    stride = tuple(stride)
+    padding = tuple(padding)
+    dilation = tuple(dilation)
+    output_padding = tuple(output_padding)
+    assert isinstance(groups, int)
+    kwargs = {
+        "stride": stride,
+        "padding": padding,
+        "dilation": dilation,
+        "transposed": transposed,
+        "output_padding": output_padding,
+        "groups": groups,
+    }
+
+    if len(x.get_size()) == len(weight.get_size()) - 1:
+        # add batch dimension to simplify rest of function
+        return L[aten.squeeze](
+            convolution(L[aten.expand](x, [1, *x.get_size()]), weight, bias, **kwargs),
+            dim=0,
+        )
+
+    out_chan, in_chan, *kernel_shape = V.graph.sizevars.guard_static_shapes(
+        weight.get_size()
+    )
+    ndim = len(kernel_shape)
+    assert ndim == len(stride) == len(padding) == len(dilation) == len(output_padding)
+
+    if (
+        config.conv_1x1_as_mm
+        and is_ones(kernel_shape)
+        and is_ones(stride)
+        and is_zeros(padding)
+        and is_ones(dilation)
+        and not transposed
+        and is_zeros(output_padding)
+        and groups == 1
+    ):
+        return convert_1x1_conv_to_mm(x, weight, bias)
+
+    if bias is not None and ir.get_device_type(x) != "cpu":
+        # peel off the bias, cudnn is slower with it
+        result = convolution(x, weight, None, **kwargs)
+        return L[aten.add](
+            result, L[aten.view](bias, [result.get_size()[1]] + ndim * [1])
+        )
+
+    x.realize()
+    weight.realize()
+    layout = conv_layout(x, weight, None, **kwargs)
+    req_stride_order = ir.get_stride_order(V.graph.sizevars.size_hints(layout.stride))
+    x = ir.ExternKernel.require_stride_order(x, req_stride_order)
+    weight = ir.ExternKernel.require_stride_order(weight, req_stride_order)
+
+    if bias is None:
+        args = (x, weight)
+        kwargs["bias"] = None
+    else:
+        args = (x, weight, bias)
+        bias.realize()
+        bias.freeze_layout()
+        V.graph.sizevars.guard_static_shapes(bias.get_size())
+
+    choices = [aten_convolution.bind(args, layout, **kwargs)]
+    if (
+        use_triton_template(layout)
+        # templates only support these:
+        and ndim == 2
+        and is_ones(dilation)
+        and not transposed
+        and is_zeros(output_padding)
+        # there are some odd models where this check fails (e.g. shufflenet_v2_x1_0)
+        and V.graph.sizevars.maybe_guard_equals(in_chan, x.get_size()[1])
+    ):
+        if (
+            is_ones(kernel_shape)
+            and is_ones(stride)
+            and is_zeros(padding)
+            and groups == 1
+        ):
+            choices.append(aten_conv1x1_via_mm.bind(args, layout))
```

Comment:
where is bias addition handled for this one (and for triton templates below)?

Reply:
We split off the bias into another op on line 312.

Reply:
For cuda, shouldn't we add addmm to 1x1 options with bias?

Reply:
Yeah, I think that would be a good option to add.  I also want to add alternate layout options.